packages/common/src/TransactionUtils.ts

```diff
@@ -101,17 +132,32 @@ export const runTransaction = async ({
     const maximumGasPriceMultiple = 2 * 3;
     // Pre-London transactions require `gasPrice`, London transactions require `maxFeePerGas` and `maxPriorityFeePerGas`
 
-    let receipt;
+    let receipt: any;
 
     // If the config contains maxPriorityFeePerGas then this is a London transaction. In this case, simply use the
     // provided config settings but double the maxFeePerGas to ensure the transaction is included, even if the base fee
     // spikes up. The difference between the realized base fee and maxFeePerGas is refunded in a London transaction.
     if (transactionConfig.maxFeePerGas && transactionConfig.maxPriorityFeePerGas) {
-      receipt = await transaction.send({
-        ...transactionConfig,
-        maxFeePerGas: parseInt(transactionConfig.maxFeePerGas.toString()) * 2,
-        type: "0x2",
-      } as SendOptions);
+      // If waitForMine is set (default) then code blocks until the transaction is mined and a receipt is returned.
+      if (waitForMine)
+        receipt = await transaction.send({
+          ...transactionConfig,
+          maxFeePerGas: parseInt(transactionConfig.maxFeePerGas.toString()) * 2,
+          type: "0x2",
+        } as SendOptions);
+      // Else, waitForMine is false and we return the transaction hash immediately as soon as it is included in the
+      // mempool. Receipt is a promise of the pending transaction that can be awaited later to ensure block inclusion.
+      else {
+        const receiptPromiseEvent = transaction.send({
+          ...transactionConfig,
+          maxFeePerGas: parseInt(transactionConfig.maxFeePerGas.toString()) * 2,
+          type: "0x2",
+        } as SendOptions);
+        const transactionHash: string = await new Promise((resolve) => {
+          receiptPromiseEvent.on("transactionHash", (transactionHash) => resolve(transactionHash));
+        });
+        return { receipt: receiptPromiseEvent as any, transactionHash, returnValue, transactionConfig };
```

Comment:
Any reason we can't just set the receipt as we do in the if above and use the existing return statement below?

Reply:
Any thoughts on this comment?

Reply:
main reason is we get the transaction hash in a differnt way in this code block than in the normal send logic (i.e we need to wait on the `transactionHash` promise resolution, and cant pull from the `receipt`).

I'm going to refactor this to have a `transactionhash` defined above so we can have a conistant return flow.

Reply:
> I'm going to refactor this to have a transactionhash defined above so we can have a conistant return flow.

agreed with this